backend/module/chaos/serverexperimentation/xds/ecds.go

```diff
@@ -0,0 +1,255 @@
+
+	faultFilterTypeURL = `type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault`
+
+	delayPercentRuntime    = `ecds_runtime_override_do_not_use.http.delay.percentage`
+	delayDurationRuntime   = `ecds_runtime_override_do_not_use.http.delay.fixed_duration_ms`
+	abortHttpStatusRuntime = `ecds_runtime_override_do_not_use.http.abort.http_status`
+	abortPercentRuntime    = `ecds_runtime_override_do_not_use.http.abort.abort_percent`
+)
+
+// Default abort and delay fault configs
+var DefaultAbortFaultConfig = &gcpFilterFault.FaultAbort{
+	ErrorType: &gcpFilterFault.FaultAbort_HttpStatus{
+		HttpStatus: 503,
+	},
+	Percentage: &gcpType.FractionalPercent{
+		Numerator:   0,
+		Denominator: gcpType.FractionalPercent_HUNDRED,
+	},
+}
+
+var DefaultDelayFaultConfig = &gcpFilterCommon.FaultDelay{
+	FaultDelaySecifier: &gcpFilterCommon.FaultDelay_FixedDelay{
+		FixedDelay: &duration.Duration{
+			Nanos: 1000000, // 0.001 second
+		},
+	},
+	Percentage: &gcpType.FractionalPercent{
+		Numerator:   0,
+		Denominator: gcpType.FractionalPercent_HUNDRED,
+	},
+}
+
+type SafeEcdsResourceMap struct {
+	mu                    sync.Mutex
+	requestedResourcesMap map[string][]string
+}
+
+type ECDSConfig struct {
+	enabledClusters map[string]struct{}
+
+	ecdsResourceMap *SafeEcdsResourceMap
+}
+
+func generateECDSResource(experiments []*experimentation.Experiment, ttl *time.Duration, logger *zap.SugaredLogger) []gcpTypes.ResourceWithTtl {
+	var downstreamCluster string
+	var upstreamCluster string
+	var isExternalFault = false
+
+	abort := DefaultAbortFaultConfig
+	delay := DefaultDelayFaultConfig
+
+	for _, experiment := range experiments {
+		httpFaultConfig := &serverexperimentation.HTTPFaultConfig{}
+		if !maybeUnmarshalFaultTest(experiment, httpFaultConfig) {
+			continue
+		}
+
+		var err error
+		upstreamCluster, downstreamCluster, err = getClusterPair(httpFaultConfig)
+		if err != nil {
+			logger.Errorw("Invalid http fault config", "config", httpFaultConfig)
+			continue
+		}
+
+		isExternalFault, abort, delay, err = createAbortDelayConfig(httpFaultConfig, downstreamCluster)
+		if err != nil {
+			logger.Errorw("Unable to create ECDS filter fault", "config", httpFaultConfig)
+			continue
+		}
+
+		// We can only perform one fault test per cluster with ECDS. Hence we break.
+		break
+	}
+
+	faultFilter := &gcpFilterFault.HTTPFault{
+		Delay: delay,
+		Abort: abort,
+
+		// override runtimes so that default runtime is not used.
+		DelayPercentRuntime:    delayPercentRuntime,
+		DelayDurationRuntime:   delayDurationRuntime,
+		AbortHttpStatusRuntime: abortHttpStatusRuntime,
+		AbortPercentRuntime:    abortPercentRuntime,
+	}
+
+	var faultFilterName string
+	if isExternalFault {
+		faultFilterName = fmt.Sprintf(FaultFilterConfigNameForExternalFault, upstreamCluster)
+	} else {
+		faultFilterName = FaultFilterConfigNameForInternalFault
+
+		// match downstream cluster with request header else fault will be applied to requests coming from all downstream
```

Comment:
This is a bit hard to parse if you don't already know what's going on, how about something like `We match against the EnvoyDownstreamServiceCluster header to only apply these faults to the specific downstream cluster`?